server/pages/index.js

```diff
@@ -374,22 +377,57 @@ module.exports = function() {
 	} );
 
 	if ( config.isEnabled( 'manage/themes/details' ) ) {
+		function updateRenderCache( themeSlug ) {
+			let themeData = themeDetails.get( themeSlug );
+			wpcom.undocumented().themeDetails( themeSlug, ( error, data ) => {
+				if ( error ) {
+					console.log( `Error fetching theme ${ themeSlug } details: `, error.message || error );
+				} else if ( ! themeDetails.has( themeSlug ) || ( Date( data.date_updated ) > Date( themeData.date_updated ) ) ) {
+					console.log( 'caching', themeSlug );
+					themeDetails.set( themeSlug, data );
+					renderThemeSheet( data );
+				}
+			} );
+		}
 
-		app.get( '/themes/:theme_slug', function( req, res ) {
+		function renderThemeSheet( theme ) {
+			const context = {};
+			const store = createReduxStore();
+			store.dispatch( {
+				type: ThemesActionTypes.RECEIVE_THEME_DETAILS,
+				themeId: theme.id,
+				themeName: theme.name,
+				themeAuthor: theme.author,
+				themeScreenshot: theme.screenshot,
+				themePrice: theme.price ? theme.price.display : undefined,
+				themeDescription: theme.description,
+				themeDescriptionLong: theme.description_long,
+				themeSupportDocumentation: theme.extended ? theme.extended.support_documentation : undefined,
+			} );
+
+			store.dispatch( setSection( 'themes', { hasSidebar: false, isFullScreen: true } ) );
+			context.initialReduxState = pick( store.getState(), 'ui', 'themes' );
+
+			Object.assign( context, render( (
+				<ReduxProvider store={ store }>
+					<LayoutLoggedOutDesign store={ store } routeName={ 'themes' } match={ { theme_slug: theme.id } } />
+				</ReduxProvider>
+			) ) ); // FIXME(ehg): we probably want to get rid of memoize, and use a straight up cache in the render module
+
+			return context;
+		};
+
+		app.get( '/themes/:theme_slug/:section?', function( req, res ) {
 			const context = getDefaultContext( req );
-
 			if ( config.isEnabled( 'server-side-rendering' ) ) {
-				i18n.initialize();
-				const store = createReduxStore();
-
-				store.dispatch( setSection( 'themes', { hasSidebar: false, isFullScreen: true } ) );
-				context.initialReduxState = pick( store.getState(), 'ui' );
+				const theme = themeDetails.get( req.params.theme_slug );
+				if ( theme ) {
+					Object.assign( context, renderThemeSheet( theme ) );
+					console.log( 'found theme!', theme.id );
+				}
 
-				Object.assign( context, render( (
-					<ReduxProvider store={ store }>
-						<LayoutLoggedOutDesign store={ store } routeName={ 'themes' } match={ { theme_slug: req.params.theme_slug } } />
-					</ReduxProvider>
-				) ) );
+				i18n.initialize();
+				req.params.theme_slug && updateRenderCache( req.params.theme_slug ); // TODO(ehg): We don't want to hit the endpoint for every req. Debounce based on theme arg?
```

Comment:
More events in the loop? Maybe it's a pre-optimisation, although I wonder if we rate limit the API on the internal network.